### ISSUE TITLE
Support private dependencies

### ISSUE THREAD

**Issue by triblondon** [polyfillpolyfill/polyfill-service#280]
In #254, @jonathantneal makes the point that while our current `Element.prototype.mutation` polyfill would benefit from being split up into separate features for replace, append, prepend, before and after, all of these would require exactly the same private mutation code.

Therefore it may be worth considering a mechanism by which functions that are not standardised web APIs can be treated as separate features by the service on which real features can depend.

An easy way to do this would be to:
1. Wrap the entire polyfill output in a closure and define a `PolyfillService` variable at the top of it, initialised to `{shared:{}}`.
2. Create private features by using the `PolyfillService.shared` namespace
3. In the polyfill.js for the private feature, assign the exported function to `PolyfillService`, eg `PolyfillService.shared.mutation = function(...) { ... };`
4. In the config.js for the real feature, add the private feature to the normal dependency list
5. In the polyfill.js for the real feature, use `PolyfillService.shared.mutation`

In terms of the service functionality, the only change needed here is the outer wrapper closure, which is only needed if you want to avoid polluting the global scope with the `PolyfillService` object.

Jonathan has a slightly different proposal in #254.

**Comment by jonathantneal** [comment]
Initial thoughts:
- We should avoid polluting the global scope.
- An automated task can assemble closures / Polyfills should be closure-free.
- We could use `config.json` or create a `closure.js` file to assign private methods and variables.

If the polyfill service wraps every response in a closure, then this becomes even easier.

**Comment by triblondon** [comment]
@jonathantneal OK, so are you satisfied with the approach I outlined in the original post above?  So there's one big closure around the whole polyfill bundle, rather than individual closures around each polyfill.  Dependencies can just be named clearly using a prefix: `PolyfillService.shared.thing`, and listed as normal dependencies - don't see a strong argument for a new config property, and if we use dependencies it will just flow through the existing dep resolution logic.

**Comment by triblondon** [comment]
Just to be clear, the `PolyfillService` variable lives inside the outer closure, so is not a global.

**Comment by jonathantneal** [comment]
That’s great. Could you explain the reason behind the `PolyfillService` object?

We have an example of the wrapped closure in [Array.isArray](https://github.com/Financial-Times/polyfill-service/blob/master/polyfills/Array.isArray/polyfill.js).

``` js
(function (toString) {
    Object.defineProperty(Array, 'isArray', {
        configurable: true,
        value: function isArray(object) {
            return toString.call(object) === '[object Array]';
        },
        writable: true
    });
})(Object.prototype.toString);
```

With one great big closure, we could now do something like:

``` js
var ObjectToString = Object.prototype.toString;

Object.defineProperty(Array, 'isArray', {
    configurable: true,
    value: function isArray(object) {
        return ObjectToString.call(object) === '[object Array]';
    },
    writable: true
});
```

I figure JavaScript compressor can go in later and reduce these declarations for the minified version if they occur multiple times (I could imagine this for `Element.prototype`).

**Comment by triblondon** [comment]
So to take your example, I'm not sure why you wouldn't simply use Object.prototype.toString in the Array.isArray code:

``` javascript
Object.defineProperty(Array, 'isArray', {
    configurable: true,
    value: function isArray(object) {
        return Object.prototype.toString.call(object) === '[object Array]';
    },
    writable: true
});
```

But where the function isn't already a native global (or a property of a native global), and we want to share it, that's where the shared object would be used:

``` javascript
// Wrapping closure
(function () {
    var PolyfillService = {shared:{}};

    // /polyfills/PolyfillService.shared.enqueueMicrotask/polyfill.js
    PolyfillService.shared.enqueueMicrotask = function(task) { ... };

    // /polyfills/Promise/polyfill.js
    self.Promise = function() {
        PolyfillService.shared.enqueueMicrotask(foo);
    }
}());
```

I added the `shared` subobject in case we later want to extend at the PolyfillService level, and felt it was probably worth being clear over being concise.  If you feel that length is excessive we could drop the grouping object entirely and just use a naming convention, like an underscore:

``` javascript
// Wrapping closure
(function () {

    // /polyfills/_enqueueMicrotask/polyfill.js
    var _enqueueMicrotask = function(task) { ... };

    // /polyfills/Promise/polyfill.js
    self.Promise = function() {
        _enqueueMicrotask(foo);
    }
}());
```

In fact, having done all that I prefer the second form - I'm leaning more towards thinking that a grouping namespace is unnecessary.

**Comment by jonathantneal** [comment]
> I'm not sure why you wouldn't simply use Object.prototype.toString in the Array.isArray code:

We do it for performance. Using local variables to cache deep properties prevents class checks. Whether this is needed in modern browsers is another question.

> I'm leaning more towards thinking that a grouping namespace is unnecessary.

I agree! An underscore would be familiar to developers (sorry, [Papa Crockford](http://jslinterrors.com/unexpected-dangling-_-in-a)).

**Comment by triblondon** [comment]
OK, so, changes to the service:
- Exclude any features prefixed with `_` from the feature list in the documentation
- Surround the entire polyfill bundle with `(function() { ... }());`

And, er, I think that's it!  In the meantime, you could modify your mutation changes to work as we've agreed here.  But this won't take long...

**Comment by jonathantneal** [comment]
I’ll do that now.

mutation => _mutation
enqueueMicrotask => _enqueueMicrotask

I’ll also check the other polyfills to make sure they attach themselves to the global `this`. If they don’t already, they should.

**Comment by triblondon** [comment]
Actually, I'll make the service change in your `enqueueMicrotask` branch, then I'll be able to test it once you've normalised to the naming convention agreed above.

**Comment by jonathantneal** [comment]
The variables and references have been updated in both branches.

**Comment by triblondon** [comment]
Implemented in #287
